### PR TITLE
Kubernetes Pass Through API & SDK Initialization

### PR DESCRIPTION
**Pull request by andrewballantyne**
Addresses #259 
https://issues.redhat.com/browse/RHODS-4575

Remaining Tasks:
- [x] Migrate a proper example of the backend over to use the k8s utilities
- [x] Test all of the k8s resource utilities on a couple resource types; eg:
    - [x] ConfigMaps
    - [x] Secrets
    - [x] PVCs

Known limitations today:
- The SDK has a url crafting issue with k8sCreateResource which does not call kube cleanly (Bug: https://issues.redhat.com/browse/HAC-1825) -- this is handled on the server with a URL manipulation
- The SDK crud utilities suppress failed errors from being bubbled back to the caller  (Bug: https://issues.redhat.com/browse/HAC-2152) -- no real workaround for this one

### CODE REVIEW

File: backend/src/routes/api/k8s/pass-through.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    // Core SDK builds the wrong path for k8s -- can't post to a resource name; remove the name from the url
+    // eg: POST /.../configmaps/my-config-map => POST /.../configmaps
+    const urlParts = url.split('/');
+    urlParts.pop();
+    safeURL = urlParts.join('/');
+  }
+
+  return new Promise((resolve, reject) => {
+    const kubeOptions: Parameters<typeof kc.applyToRequest>[0] = { url: safeURL };
+    kc.applyToRequest(kubeOptions).then(() => {
```
Comment (andrewballantyne) [inline backend/src/routes/api/k8s/pass-through.ts:31]: This needs to be a bearer token from the user (from the oauth proxy). This is going to be the same service account that started the backend.

Comment (andrewballantyne) [reply backend/src/routes/api/k8s/pass-through.ts:31]: Talking with @lucferbux -- we need to write a fastify plugin to be `dev.ts` (like `kube.ts`) that will just manage pulling the current logged in user (from `kube.ts`) and replace the `x-forwarded-...` header (it likely won't exist, but if it does toss it out and replace it). 

That way when we make local development calls the same way we would do on-cluster as the service account used to do. `x-forwarded-...` header will have the token OF THE USER, not the service account when this PR merges (in both local & on cluster -- so it is the same experience).

**Review by dlabaj — APPROVED** [review]
/lgtm

File: backend/src/routes/api/k8s/pass-through.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  let headers;
+  if (DEV_MODE) {
+    // In dev mode, we always are logged in fully -- no service accounts
+    headers = kubeHeaders;
+  } else {
+    // When not in dev mode, we want to switch the token from the service account to the user
+    const accessToken = request.headers[USER_ACCESS_TOKEN];
+    headers = {
+      ...kubeHeaders,
+      Authorization: `Bearer ${accessToken}`,
```
Comment (andrewballantyne) [inline backend/src/routes/api/k8s/pass-through.ts:47]: @lucferbux How's this logic?

File: frontend/src/services/configMapService.ts [hunk 3]
```diff
@@ -1,33 +1,33 @@
...
+  k8sUpdateResource,
+} from '@openshift/dynamic-plugin-sdk-utils';
+import { ConfigMapModel } from '../models';
 import { ConfigMap, DeleteStatus } from '../types';
 
 export const getConfigMap = (projectName: string, configMapName: string): Promise<ConfigMap> => {
-  const url = `/api/configmaps/${projectName}/${configMapName}`;
-  return axios.get(url).then((response) => {
-    return response.data;
+  return k8sGetResource<ConfigMap>({
```
Comment (DaoDaoNoCode) [inline frontend/src/services/configMapService.ts:11]: Maybe we can get rid of the braces and `return` if we are not going to do any other operation before the return statement?

Comment (andrewballantyne) [reply frontend/src/services/configMapService.ts:11]: I think if we want this kinda style, we should just force it in the linter. This is a preference thing of code style and we could probably disagree until we go blue in the face lol. Let us talk about linter improvements. 

I logged https://github.com/opendatahub-io/odh-dashboard/issues/469 to have that conversation.

**Comment by andrewballantyne** [comment]
Testing this on cluster ran me across odd oauth x-forwarded token issues. My `kube:admin` user as well as a cluster-admin user both got 403 forbidden on requesting a secret or configmap that was not even there.

This PR will continue to be on hold until more investigation is done.

**Comment by samuelvl** [comment]
The current oauth scope is `user:info user:check-access` and it should be `user:full` but using a service account as an oauth client [doesn't allow this](https://docs.openshift.com/container-platform/4.11/authentication/using-service-accounts-as-oauth-client.html#service-accounts-as-oauth-clients_using-service-accounts-as-oauth-client).

We have to create a dedicated `OAuthClient` object:

```yaml
kind: OAuthClient
apiVersion: oauth.openshift.io/v1
metadata:
  name: odh-dashboard
  labels:
    app: odh-dashboard
secret: ${CLIENT_SECRET}
redirectURIs:
  - ${RHODS_DASHBOARD_URL}
grantMethod: auto
```

And modify the oauth proxy arguments to include it:

```yaml
args:
  ...
  - --client-id=odh-dashboard
  - --client-secret=${CLIENT_SECRET}
  - --scope=user:full
```

I'm working on a PR to make this change.

**Comment by andrewballantyne** [comment]
Samu figured out a way to get this to work using permissions settings he noted in his previous comment. There was some wonky setup issues but it looks like this is the right path. My PR will wait until we can get this setup more mainstream and less finicky.

**Comment by andrewballantyne** [comment]
/unhold

Going to actually get this merged in asap so we can start using this in the `f/dsg` branch -- the current issue on cluster is a blocker to merge the feature, but we have a longer window to address that.

**Pull request merged**